src/v/cloud_storage/recursive_directory_walker.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cloud_storage/logger.h"
+#include "seastarx.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/sstring.hh>
+
+#include <chrono>
+#include <memory>
+#include <utility>
+
+namespace cloud_storage {
+
+struct recursive_directory_walker {
+    using walker_type = std::function<ss::future<>(ss::directory_entry)>;
+
+    struct file_list_item {
+        ss::lowres_clock::time_point access_time;
+        ss::sstring filename;
+        uint64_t size;
+    };
+
+    std::deque<file_list_item> files;
```

Comment:
it's probably should be sorted by timestamp, from old to new

Reply:
Done.